codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,469 @@
+            streamInfo.inStreamsToFireChildReadComplete = true;
+        }
+    }
+
+    final class ConnectionListener extends Http2ConnectionAdapter {
+        @Override
+        public void onStreamActive(Http2Stream stream) {
+            if (ctx == null) {
+                // UPGRADE stream is active before handlerAdded().
+                return;
+            }
+            // If it is an outgoing stream, then we already created the channel.
+            // TODO: support outgoing streams
+            if (stream.getProperty(streamInfoKey) != null) {
+                return;
+            }
+            ChannelFuture future = createStreamChannel(ctx, stream, streamHandler);
+            stream.setProperty(streamInfoKey, new StreamInfo((Http2StreamChannel) future.channel()));
+        }
+
+        @Override
+        public void onStreamClosed(Http2Stream stream) {
+            final StreamInfo streamInfo = (StreamInfo) stream.removeProperty(streamInfoKey);
+            if (streamInfo != null) {
+                final EventLoop eventLoop = streamInfo.childChannel.eventLoop();
+                if (eventLoop.inEventLoop()) {
+                    onStreamClosed0(streamInfo);
+                } else {
+                    eventLoop.execute(new OneTimeTask() {
+                        @Override
+                        public void run() {
+                            onStreamClosed0(streamInfo);
+                        }
+                    });
+                }
+            }
+        }
+
+        private void onStreamClosed0(StreamInfo streamInfo) {
+            streamInfo.childChannel.onStreamClosedFired = true;
+            streamInfo.childChannel.fireChildRead(AbstractHttp2StreamChannel.CLOSE_MESSAGE);
+        }
+
+        @Override
+        public void onGoAwayReceived(final int lastStreamId, long errorCode, ByteBuf debugData) {
+            final Http2GoAwayFrame goAway = new DefaultHttp2GoAwayFrame(errorCode, debugData);
+            try {
+                http2Handler.connection().forEachActiveStream(new Http2StreamVisitor() {
+                    @Override
+                    public boolean visit(Http2Stream stream) {
+                        if (stream.id() > lastStreamId
+                                && http2Handler.connection().local().isValidStreamId(stream.id())) {
+                            StreamInfo streamInfo = (StreamInfo) stream.getProperty(streamInfoKey);
+                            // TODO: Can we force a user interaction pattern that doesn't require us to duplicate()?
+                            streamInfo.childChannel.pipeline().fireUserEventTriggered(goAway.duplicate().retain());
+                        }
+                        return true;
+                    }
+                });
+            } catch (Throwable t) {
+                invoker(ctx).invokeExceptionCaught(ctx, t);
+            }
+            ctx.fireUserEventTriggered(goAway.duplicate().retain());
```

Comment:
This callback method (`onGoAwayReceived`) isn't given ownership of the buffer, only temporary access. https://github.com/netty/netty/blob/72accceeacf15d13358d5c46d5c414fc49f6c38d/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java#L158